common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -316,22 +355,39 @@ public String[] getMergedBlockDirs(String appId) {
   @Override
   public void applicationRemoved(String appId, boolean cleanupLocalDirs) {
     logger.info("Application {} removed, cleanupLocalDirs = {}", appId, cleanupLocalDirs);
-    AppShuffleInfo appShuffleInfo = appsShuffleInfo.remove(appId);
+    // Cleanup the DB within critical section to gain the consistency between
+    // DB and in-memory hashmap.
+    AtomicReference<AppShuffleInfo> ref = new AtomicReference<>(null);
+    appsShuffleInfo.compute(appId, (id, info) -> {
+      if (null != info) {
+        // Try cleaning up this application attempt local paths information
+        // and also the local paths information from former attempts in DB.
+        removeAppAttemptPathInfoFromDB(info.appId, info.attemptId);
+        if (info.attemptId != UNDEFINED_ATTEMPT_ID) {
+          for (int formerAttemptId = info.attemptId - 1; formerAttemptId >= 0; formerAttemptId--) {
+            removeAppAttemptPathInfoFromDB(info.appId, formerAttemptId);
+          }
+        }
```

Comment:
nit: Reformulate it as 

```suggestion
        if (info.attemptId != UNDEFINED_ATTEMPT_ID) {
          for (int formerAttemptId = 0; formerAttemptId < info.attemptId; formerAttemptId ++) {
            removeAppAttemptPathInfoFromDB(info.appId, formerAttemptId);
          }
        }
```

We should probably pull this into a separate private method - something like `removeOldApplicationAttemptsFromDb(info)`

Reply:
Updated